demos/defender/defender_demo/defender_demo.c

```diff
@@ -0,0 +1,573 @@
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/* Standard includes. */
+#include <stdlib.h>
+
+/* POSIX includes. */
+#include <unistd.h>
+
+/* Demo config. */
+#include "demo_config.h"
+
+/* Metrics collector. */
+#include "metrics_collector.h"
+
+/* Report builder. */
+#include "report_builder.h"
+
+/* MQTT operations. */
+#include "mqtt_operations.h"
+
+/* JSON Library. */
+#include "core_json.h"
+
+/* Device Defender Client Library. */
+#include "defender.h"
+
+/**
+ * @brief Status values of the device defender report.
+ */
+typedef enum
+{
+    ReportStatusNotReceived,
+    ReportStatusAccepted,
+    ReportStatusRejected
+} ReportStatus_t;
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Network Stats.
+ */
+static NetworkStats_t networkStats;
+
+/**
+ * @brief Open TCP ports array.
+ */
+static uint16_t openTcpPorts[ OPEN_TCP_PORTS_ARRAY_SIZE ];
+
+/**
+ * @brief Open UDP ports array.
+ */
+static uint16_t openUdpPorts[ OPEN_UDP_PORTS_ARRAY_SIZE ];
+
+/**
+ * @brief Established connections array.
+ */
+static Connection_t establishedConnections[ ESTABLISHED_CONNECTIONS_ARRAY_SIZE ];
+
+/**
+ * @brief All the metrics sent in the device defender report.
+ */
+static ReportMetrics_t deviceMetrics;
+
+/**
+ * @brief Report status.
+ */
+static ReportStatus_t reportStatus;
+
+/**
+ * @brief Buffer for generating the device defender report.
+ */
+static char deviceMetricsJsonReport[ DEVICE_METRICS_REPORT_BUFFER_SIZE ];
+
+/**
+ * @brief Report Id sent in the defender report.
+ */
+static uint32_t reportId = 0;
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Callback to receive the incoming publish messages from the MQTT broker.
+ *
+ * @param[in] pPublishInfo Pointer to publish info of the incoming publish.
+ * @param[in] packetIdentifier Packet identifier of the incoming publish.
+ */
+static void publishCallback( MQTTPublishInfo_t * pPublishInfo,
+                             uint16_t packetIdentifier );
+
+/**
+ * @brief Collect all the metrics to be sent in the device defender report.
+ *
+ * @return EXIT_SUCCESS if all the metrics are successfully collected;
+ * EXIT_FAILURE otherwise.
+ */
+static int collectDeviceMetrics( void );
```

Comment:
Can we return bool instead of exit success and failure like the job demo?

Reply:
Done.